bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/internal/ui/navigator/CSymbolsContentProvider.java

```diff
@@ -154,13 +157,26 @@ protected void addChildren(Object parent, Object[] children, IProgressMonitor mo
 						if (children.length != 1 || children[0] != ERROR_ELEMENT) {
 							viewer.remove(ERROR_ELEMENT);
 						}
-						viewer.add(parent, children);
+						var limitedChildren = children;
+						var viewLimit = getViewLimit();
+						if (children.length > viewLimit && viewLimit > 0) {
+							// limit children to prevent long UI freezes...
+							// more children can be shown in navigator Treeview via 'Show next x items from remaining y'
+							// at the end of the list
+							limitedChildren = Arrays.copyOfRange(children, 0, viewLimit);
+						}
+						viewer.add(parent, limitedChildren);
 					} finally {
 						control.setRedraw(true);
 					}
 				}
 			});
 		}
+
+		private static int getViewLimit() {
+			return WorkbenchPlugin.getDefault().getPreferenceStore()
```

Comment:
I would check if can obtain this preference via `PlatformUI#getPreferenceStore()` to avoid usage of internal API

Reply:
Unfortunately 'WorkbenchPlugin.getDefault() ' does not deliver the correct store.